File: pydockenv/commands/environment.py
import json
import os
from dataclasses import dataclass
from dataclasses import field
from typing import Dict

import click
import toml

import docker
from docker.types import Mount

from pydockenv import definitions
from pydockenv.client import Client


def get_current_env():
    return os.environ.get('PYDOCKENV')


@dataclass(frozen=True)
class EnvironmentConfig:
    name: str
    python: str = 'latest'
    dependencies: Dict[str, str] = field(default_factory=dict)
    container_args: Dict[str, str] = field(default_factory=dict)
    aliases: Dict[str, Dict[str, str]] = field(default_factory=dict)

    @classmethod
    def from_file(cls, file_: str) -> 'EnvironmentConfig':
        config = toml.load(file_)['tool']['pydockenv']
        return EnvironmentConfig(**config)


def create(project_dir, file_, name, version):
    if file_:
        config = EnvironmentConfig.from_file(file_)
    else:
        config = EnvironmentConfig(name, python=version or 'latest')

    create_from_config(project_dir, config)


def create_from_config(project_dir: str, config: EnvironmentConfig):
    click.echo(f'Creating environment {config.name} with python version '
               f'{config.python}...')
    image_name = f'python:{config.python}'

    client = Client.get_instance()
    try:
        image = client.images.get(image_name)
    except docker.errors.ImageNotFound:
        click.echo(f'Image {image_name} not found, pulling...')
        image = client.images.pull('python', tag=config.python)

    create_network(config.name)
    create_env(image, project_dir, config)

    click.echo(f'Environment {config.name} with python version '
               f'{config.python} created!')


def status():
    current_env = get_current_env()
    if not current_env:
        click.echo('No active environment')
    else:
        click.echo(f'Active environment: {current_env}')


def activate(name):
    click.echo('Activating environment...')
    try:
        container = Client.get_instance().containers.get(
            definitions.CONTAINERS_PREFIX + name)
    except docker.errors.NotFound:
        click.echo(f'Environment {name} not found, exiting...')
    else:
        container.start()
        click.echo('Environment activated!')


def deactivate():
    click.echo('Deactivating current environment...')
    current_env = get_current_env()
    try:
        container = Client.get_instance().containers.get(
            definitions.CONTAINERS_PREFIX + current_env)
    except docker.errors.ImageNotFound:
        click.echo(f'Environment {current_env} not found, exiting...')
    else:
        container.stop()
        click.echo('Environment deactivated!')


def remove(name):
    click.echo(f'Removing environment {name}...')
    try:
        container = Client.get_instance().containers.get(
            definitions.CONTAINERS_PREFIX + name)
    except docker.errors.NotFound:
        click.echo(f'Environment {name} not found, exiting...')
        raise

    kwargs = {
        'force': True,
    }
    container.remove(**kwargs)
    delete_network(name)
    click.echo(f'Environment {name} removed!')


def list_environments():
    click.echo(f'Listing environments...')
    kwargs = {
        'all': True,
    }
    containers = Client.get_instance().containers.list(kwargs)

    current_env = get_current_env()
    envs = []
    for c in containers:
        if not c.name.startswith(definitions.CONTAINERS_PREFIX):
            continue

        env_name = c.name[len(definitions.CONTAINERS_PREFIX):]
        prefix = '* ' if env_name == current_env else '  '
        envs.append(f'{prefix}{env_name}')

    click.echo('\n'.join(envs))
    click.echo(f'Environments listed!')


def create_network(env_name):
    network_name = definitions.CONTAINERS_PREFIX + env_name + '_network'
    Client.get_instance().networks.create(network_name, check_duplicate=True)


def delete_network(env_name):
    network_name = definitions.CONTAINERS_PREFIX + env_name + '_network'
    try:
        network = Client.get_instance().networks.get(network_name)
    except docker.errors.ImageNotFound:
        click.echo(f'Network {network_name} not found, exiting...')
        raise

    for c in network.containers:
        network.disconnect(c)

    network.remove()


def create_env(image, project_dir, config):
    workdir = os.path.abspath(project_dir)
    mounts = [
        Mount('/usr/src', workdir, type='bind')
    ]
    kwargs = {
        'command': '/bin/sh',
        'stdin_open': True,
        'labels': {
            'workdir': workdir,
            'env_name': config.name,
            'aliases': json.dumps(config.aliases),
        },
        'name': definitions.CONTAINERS_PREFIX + config.name,
        'mounts': mounts,
        'network': definitions.CONTAINERS_PREFIX + config.name + '_network',
    }

    filtered_container_args = {k: v for k, v in config.container_args.items()
                               if k not in kwargs}
    kwargs.update(filtered_container_args)

    container = Client.get_instance().containers.create(image, **kwargs)

    if config.dependencies:
        # TODO: Remove this from here just to avoid circular imports
        from pydockenv.commands import dependency

        container.start()

        click.echo(f'Installing {len(config.dependencies)} dependencies...')
        packages = [f'{dep}{v}' for dep, v in config.dependencies.items()]
        click.echo(f'Installing {packages}...')
        dependency.install_for_container(container, packages, None)

        container.stop()

File: internal/environment/environment.go
package environment

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"

	"github.com/docker/docker/api/types"
	"github.com/docker/docker/api/types/container"
	"github.com/docker/docker/api/types/mount"
	"github.com/docker/docker/api/types/network"
	"github.com/docker/docker/client"
	"github.com/docker/docker/pkg/jsonmessage"
	"github.com/docker/docker/pkg/term"
	"github.com/se7entyse7en/pydockenv/internal/utils"
	"github.com/se7entyse7en/pydockenv/log"
	"github.com/sirupsen/logrus"
)

type Config struct {
	Name          string
	Python        string
	ProjectDir    string
	Dependencies  map[string]string
	ContainerArgs map[string]string
	Aliases       map[string]struct {
		Cmd   string
		Ports []int
	}
}

type RawConfig struct {
	Tool struct {
		Pydockenv *Config
	}
}

func Create(conf *Config) error {
	logger := log.Logger
	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return err
	}

	contImg := fmt.Sprintf("python:%s", conf.Python)
	images, err := listDockerImages(cli)
	if err != nil {

		return fmt.Errorf("cannot list docker images: %w", err)
	}

	ctxLogger := logger.WithFields(logrus.Fields{
		"image": contImg,
	})

	if _, ok := images[contImg]; !ok {
		ctxLogger.Debug("Image not found, pulling...")
		out, err := cli.ImagePull(context.Background(), contImg,
			types.ImagePullOptions{})
		if err != nil {
			return fmt.Errorf("cannot pull image: %w", err)
		}

		defer out.Close()

		termFd, isTerm := term.GetFdInfo(os.Stderr)
		jsonmessage.DisplayJSONMessagesStream(out, os.Stderr, termFd, isTerm, nil)

		ctxLogger.Debug("Image pulled!")
	}

	netName := fmt.Sprintf("%s_%s_network", utils.RESOURCES_PREFIX, conf.Name)
	ctxLogger = logger.WithFields(logrus.Fields{
		"network-name": netName,
	})

	ctxLogger.Debug("Creating network...")
	_, err = cli.NetworkCreate(
		context.Background(),
		netName,
		types.NetworkCreate{CheckDuplicate: true},
	)
	if err != nil {
		return fmt.Errorf("cannot create network: %w", err)
	}

	ctxLogger.Debug("Network created!")

	contName := fmt.Sprintf("%s_%s", utils.RESOURCES_PREFIX, conf.Name)
	workdir, err := filepath.Abs(conf.ProjectDir)
	if err != nil {
		return err
	}

	ctxLogger = logger.WithFields(logrus.Fields{
		"container-image": contImg,
		"container-name":  contName,
		"workdir":         workdir,
	})
	ctxLogger.Debug("Creating container...")

	jsonAliases, err := json.Marshal(conf.Aliases)
	if err != nil {
		return err
	}

	_, err = cli.ContainerCreate(
		context.Background(),
		&container.Config{
			Image:     contImg,
			Cmd:       []string{"/bin/sh"},
			OpenStdin: true,
			Labels: map[string]string{
				"workdir":  workdir,
				"env_name": conf.Name,
				"aliases":  string(jsonAliases),
			},
		},
		&container.HostConfig{
			NetworkMode: container.NetworkMode(netName),
			Mounts: []mount.Mount{
				{
					Source: workdir,
					Target: "/usr/src",
					Type:   mount.TypeBind,
				},
			},
		},
		&network.NetworkingConfig{},
		contName,
	)
	if err != nil {
		return fmt.Errorf("cannot create container: %w", err)
	}

	ctxLogger.Debug("Container created!")

	return nil
}

func Status() error {
	logger := log.Logger
	envName := utils.GetCurrentEnv()
	if envName == "" {
		logger.Info("No active environment")
		return nil
	}

	logger.Infof("Active environment: %s", envName)
	return nil
}

func Activate(envName string) error {
	logger := log.Logger
	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return err
	}

	contName := fmt.Sprintf("%s_%s", utils.RESOURCES_PREFIX, envName)

	ctxLogger := logger.WithFields(logrus.Fields{
		"container-name": contName,
	})
	ctxLogger.Debug("Starting container...")

	err = cli.ContainerStart(context.Background(), contName,
		types.ContainerStartOptions{})
	if err != nil {
		return fmt.Errorf("cannot start container: %w", err)
	}

	ctxLogger.Debug("Container started!")

	return nil
}

func Deactivate() error {
	logger := log.Logger

	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return err
	}

	envName := utils.GetCurrentEnv()
	contName := fmt.Sprintf("%s_%s", utils.RESOURCES_PREFIX, envName)

	ctxLogger := logger.WithFields(logrus.Fields{
		"container-name": contName,
	})
	ctxLogger.Debug("Stopping container...")

	err = cli.ContainerStop(context.Background(), contName, nil)
	if err != nil {
		return fmt.Errorf("cannot stop container: %w", err)
	}

	ctxLogger.Debug("Container stopped!")

	return nil
}

func ListEnvironments() error {
	logger := log.Logger

	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return err
	}

	currentEnvName := utils.GetCurrentEnv()
	containers, err := cli.ContainerList(
		context.Background(), types.ContainerListOptions{All: true})
	if err != nil {
		return fmt.Errorf("cannot list containers: %w", err)
	}

	var msgBuilder strings.Builder
	for _, c := range containers {
		contName := c.Names[0]
		if !strings.HasPrefix(contName, utils.RESOURCES_PREFIX) {
			continue
		}

		envName := contName[len(utils.RESOURCES_PREFIX):]
		m := fmt.Sprintf("%s\n", envName)
		if envName == currentEnvName {
			m = fmt.Sprintf("* %s", m)
		}

		msgBuilder.WriteString(m)
	}

	msg := msgBuilder.String()
	if msg == "" {
		logger.Info("No environments available")
	} else {
		logger.Info(msg)
	}

	return nil
}

func Remove(envName string) error {
	logger := log.Logger
	cli, err := client.NewClientWithOpts(client.FromEnv)
	if err != nil {
		return err
	}

	contName := fmt.Sprintf("%s_%s", utils.RESOURCES_PREFIX, envName)

	ctxLogger := logger.WithFields(logrus.Fields{
		"container-name": contName,
	})
	ctxLogger.Debug("Removing container...")

	err = cli.ContainerRemove(context.Background(), contName,
		types.ContainerRemoveOptions{
			RemoveVolumes: true,
			Force:         true,
		})
	if err != nil {
		return fmt.Errorf("cannot remove container: %w", err)
	}

	ctxLogger.Debug("Container removed!")

	netName := fmt.Sprintf("%s_%s_network", utils.RESOURCES_PREFIX, envName)
	ctxLogger = logger.WithFields(logrus.Fields{
		"network-name": netName,
	})

	ctxLogger.Debug("Removing network...")
	err = cli.NetworkRemove(context.Background(), netName)
	if err != nil {
		return fmt.Errorf("cannot remove network: %w", err)
	}

	ctxLogger.Debug("Network Removed!")
	return nil
}

func listDockerImages(cli *client.Client) (map[string]struct{}, error) {
	images, err := cli.ImageList(context.Background(),
		types.ImageListOptions{All: true})

	var imagesNames map[string]struct{}
	if err != nil {
		return imagesNames, err
	}

	imagesNames = make(map[string]struct{})
	for _, im := range images {
		for _, tag := range im.RepoTags {
			if tag == "<none>:<none>" {
				continue
			}

			imagesNames[tag] = struct{}{}
		}
	}

	return imagesNames, nil
}
